src/runtime/Codecs/FunctionCodec.cs

```diff
@@ -0,0 +1,377 @@
+
+                using (Py.GIL())
+                {
+                    var pyResult = pyAction.Invoke(pyArgs);
+                    Runtime.XIncref(pyResult.Handle);
+                    object result;
+                    Converter.ToManaged(pyResult.Handle, typeof(object), out result, true);
+                    return result;
+                }
+            };
+            return (object)func;
+        }
+
+        public bool TryDecode<T>(PyObject pyObj, out T value)
+        {
+            value = default(T);
+            var tT = typeof(T);
+            if (!IsCallable(tT))
+                return false;
+
+            var numArgs = GetNumArgs(pyObj);
+
+            if (IsAction(tT))
+            {
+                object actionObj = null;
+                if (numArgs == 0)
+                {
+                    actionObj = ConvertUnaryAction(pyObj);
+                }
+                else
+                {
+                    actionObj = ConvertVariadicObjectAction(pyObj, numArgs);
+                }
+
+                value = (T)actionObj;
+                return true;
+            }
+            else if (IsFunc(tT))
+            {
+
+                object funcObj = null;
+                if (numArgs == 0)
+                {
+                    funcObj = ConvertUnaryFunc(pyObj);
+                }
+                else
+                {
+                    funcObj = ConvertVariadicObjectFunc(pyObj, numArgs);
+                }
+
+                value = (T)funcObj;
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        public static void Register()
+        {
+            PyObjectConversions.RegisterDecoder(Instance);
+            PyObjectConversions.RegisterEncoder(Instance);
+        }
+
+        public bool CanEncode(Type type)
+        {
+            return IsCallable(type);
+        }
+
+        public PyObject TryEncode(object value)
+        {
+            if (value == null) return null;
+
+            var targetType = value.GetType();
+            ICallable callable = null;
+            if (IsUnaryAction(targetType))
+            {
+                callable = new ActionWrapper(value as Action);
+
+            }
+            else if (IsVariadicObjectAction(targetType))
+            {
+                callable = new ActionWrapper(value as Action<object[]>);
+            }
+            else if (IsUnaryFunc(targetType))
+            {
+                callable = new FunctionWrapper(value as Func<object>);
+            }
+            else if (IsVariadicObjectFunc(targetType))
+            {
+                callable = new FunctionWrapper(value as Func<object[], object>);
+            }
+            else
+            {
+                throw new Exception("object cannot be encoded!");
+            }
+
+            var methodWrapper = new MethodWrapper2(callable, "RunAction", "BinaryFunc");
+
+            //TODO - lifetime??
```

Comment:
@amos402 @lostmsu @filmor my main concern with this change.  I have no idea how to manage the lifetime of the method wrapper and ICallable.

Reply:
You can implement an `ExtensionType`.

Reply:
@amos402 thank you.  I don't know what that is but I will look into it